src/uu/expr/src/syntax_tree.rs

```diff
@@ -200,6 +194,68 @@ const PRECEDENCE: &[&[(&str, BinOp)]] = &[
     &[(":", BinOp::String(StringOp::Match))],
 ];
 
+#[derive(Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub enum NumOrStr {
+    Num(BigInt),
+    Str(String),
+}
+
+impl From<usize> for NumOrStr {
+    fn from(num: usize) -> Self {
+        Self::Num(BigInt::from(num))
+    }
+}
+
+impl From<BigInt> for NumOrStr {
+    fn from(num: BigInt) -> Self {
+        Self::Num(num)
+    }
+}
+
+impl From<String> for NumOrStr {
+    fn from(str: String) -> Self {
+        Self::Str(str)
+    }
+}
+
+impl From<NumOrStr> for Option<usize> {
+    fn from(s: NumOrStr) -> Self {
+        match s.into() {
```

Comment:
This is where the conversions get confusing, I think. It's now unclear what the `.into()` is doing here exactly, because it relies a bit too much on type inference.

Reply:
Have kept the `from` trait here and changed `into` to `eval_as_bigint`.